lib/consensus/raft/raft.go

```diff
@@ -0,0 +1,718 @@
+// is wrong. This is a best-effort solution as there is no way to tell Raft
+// to stop processing entries because we want to take a snapshot before
+// shutting down.
+func (rw *raftWrapper) snapshotOnShutdown() error {
+	var err error
+	for i := 0; i < maxShutdownSnapshotRetries; i++ {
+		ctx, cancel := context.WithTimeout(context.Background(), waitForUpdatesShutdownTimeout)
+		err = rw.WaitForUpdates(ctx)
+		cancel()
+		if err != nil {
+			logger.Warn("timed out waiting for state updates before shutdown. Snapshotting may fail")
+			return rw.Snapshot()
+		}
+
+		err = rw.Snapshot()
+		if err == nil {
+			return nil // things worked
+		}
+
+		// There was an error
+		err = errors.New("could not snapshot raft: " + err.Error())
+		logger.Warnf("retrying to snapshot (%d/%d)...", i+1, maxShutdownSnapshotRetries)
+	}
+	return err
+}
+
+// Shutdown shutdown Raft and closes the BoltDB.
+func (rw *raftWrapper) Shutdown(ctx context.Context) error {
+	//_, span := trace.StartSpan(ctx, "consensus/raft/Shutdown")
+	//defer span.End()
+
+	errMsgs := ""
+
+	rw.cancel()
+
+	err := rw.snapshotOnShutdown()
+	if err != nil {
+		errMsgs += err.Error() + ".\n"
+	}
+
+	future := rw.raft.Shutdown()
+	err = future.Error()
+	if err != nil {
+		errMsgs += "could not shutdown raft: " + err.Error() + ".\n"
+	}
+
+	err = rw.boltdb.Close() // important!
+	if err != nil {
+		errMsgs += "could not close boltdb: " + err.Error()
+	}
+
+	if errMsgs != "" {
+		return errors.New(errMsgs)
+	}
+
+	return nil
+}
+
+// AddPeer adds a peer to Raft
+func (rw *raftWrapper) AddPeer(ctx context.Context, peerId peer.ID) error {
+	//ctx, span := trace.StartSpan(ctx, "consensus/raft/AddPeer")
+	//defer span.End()
+
+	// Check that we don't have it to not waste
+	// log entries if so.
+	peers, err := rw.Peers(ctx)
+	if err != nil {
+		return err
+	}
+	if find(peers, peerId.String()) {
+		logger.Infof("%s is already a raft peerStr", peerId.String())
+		return nil
+	}
+
+	err = rw.host.Connect(ctx, peer.AddrInfo{ID: peerId})
+	if err != nil {
+		return err
+	}
+
+	future := rw.raft.AddVoter(
+		hraft.ServerID(peerId.String()),
+		hraft.ServerAddress(peerId.String()),
+		0,
+		0,
+	) // TODO: Extra cfg value?
+	err = future.Error()
+	if err != nil {
+		logger.Error("raft cannot add peer: ", err)
+	}
+	return err
+}
+
+// RemovePeer removes a peer from Raft
+func (rw *raftWrapper) RemovePeer(ctx context.Context, peer string) error {
+	//ctx, span := trace.StartSpan(ctx, "consensus/RemovePeer")
+	//defer span.End()
+
+	// Check that we have it to not waste
+	// log entries if we don't.
+	peers, err := rw.Peers(ctx)
+	if err != nil {
+		return err
+	}
+	if !find(peers, peer) {
+		logger.Infof("%s is not among raft peers", peer)
+		return nil
+	}
+
+	if len(peers) == 1 && peers[0] == peer {
+		return errors.New("cannot remove ourselves from a 1-peer cluster")
+	}
+
+	rmFuture := rw.raft.RemoveServer(
+		hraft.ServerID(peer),
+		0,
+		0,
+	) // TODO: Extra cfg value?
```

Comment:
Do we know what this todo is about?

Reply:
no idea. This was copy pasta. I'll remove it

Reply:
No idea. This was copy pasta. Removed

Reply:
I also have no idea :man_shrugging: